pkg/sql/pgwire/conn.go

```diff
@@ -321,6 +321,44 @@ Loop:
 		typ, n, err = c.readBuf.ReadTypedMsg(&c.rd)
 		c.metrics.BytesInCount.Inc(int64(n))
 		if err != nil {
+			// Only perform this logic on ClientMsgSimpleQuery for v20.2.
+			// We cannot safely do the rest of this (i.e. for portals) without doing
+			// the todo message by jordan regarding error in extended protocol state
+			// below.
+			if pgwirebase.IsMessageTooBigError(err) && typ == pgwirebase.ClientMsgSimpleQuery {
+				log.VEventf(ctx, 1, "pgwire: found big error message; attempting to slurp bytes and return error: %s", err)
+
+				// Slurp the remaining bytes.
+				slurpN, slurpErr := c.readBuf.SlurpBytes(&c.rd, pgwirebase.GetMessageTooBigSize(err))
+				c.metrics.BytesInCount.Inc(int64(slurpN))
+				if slurpErr != nil {
+					log.VEventf(ctx, 1, "pgwire: error slurping remaining bytes: %s", slurpErr)
+					break Loop
+				}
+
+				// Write out the error over pgwire.
+				if err := writeErr(
+					ctx,
+					&sqlServer.GetExecutorConfig().Settings.SV,
+					err,
+					&c.msgBuilder,
+					&c.writerState.buf,
+				); err != nil {
+					log.VEventf(ctx, 1, "pgwire: error writing too big error message to the client")
+					break Loop
+				}
+
+				// We have to send the sync message back as well.
+				if err = c.stmtBuf.Push(ctx, sql.Sync{}); err != nil {
+					log.VEventf(ctx, 1, "pgwire: error writing sync to the client whilst message is too big")
+					break Loop
+				}
+
+				// We cannot break here, as more messages may come in that results in the clear
```

Comment:
Can you say more words here in the comment about how this effects the stream of messages: is the client able to recover its session no matter which packet is too large?

or is the tolerance for large packets limited to a certain message type?

I don't see this distinction made in the test:

1. if the code is meant to only tolerate a particular message type, you want a test that demonstrates that the connection is still closed for message types that are not supported

2. conversely, if the code is meant to tolerate excess type _any_ msg type, we want tests for many of them.

All in all I would recommend various combinations of tests in `pgwire_test.go` (i.e. datadriven-tests) with a small message size limit so that you can exercise this in many combinations of message types.

Reply:
> Can you say more words here in the comment about how this effects the stream of messages: is the client able to recover its session no matter which packet is too large?
> or is the tolerance for large packets limited to a certain message type?

Done.